### ISSUE TITLE
request: pub workspaces

### ISSUE THREAD

**Issue by lishaduck** [invertase/melos#747]
### Is there an existing feature request for this?

- [X] I have searched the existing issues.

### Description

Once Pub Workspaces land in stable, it would be great if Melos replaced `dependency_overrides` generation in favor of Pub Workspaces.
I believe this is planned, but I didn't see an issue.

### Reasoning

Pub Workspaces enable a single analyzer scope, which both improves performance and reduces the maintenance burden on the Melos team.

### Additional context and comments

Thanks for melos! It works great!

**Comment by spydon** [comment]
It is indeed planned, thanks for opening an issue so that we can track it!
I'll link in the Flutter design doc for it is here too: https://flutter.dev/go/pub-workspace

**Comment by spydon** [comment]
Related: https://github.com/dart-lang/pub/issues/4376

**Comment by spydon** [comment]
Related (globbing in workspace definition, which would make future migrations easier): https://github.com/dart-lang/pub/issues/4391

**Comment by Leptopoda** [comment]
I can see how globs are nice to have for this use case, but I wouldn't consider it a blocker.
I've migrated some repos to workspaces by using `melos list -rp`.

I'd love to use melos + workspaces sooner rather than waiting for this that isn't even such a big inconvenience.

**Comment by spydon** [comment]
@Leptopoda I agree, it's not a blocker, possibly we could even write a simple migration script for it.

**Comment by xaethos** [comment]
Do Melos and Pub workspaces work together right now? Has anybody gotten a repo migrated by hand?

A few weeks back I took a stab at migrating a repo and couldn't get everything to work. I cannot remember for sure, but I think my blocker ultimately was the Flutter toolchain not handling Pub workspaces, though believe some Melos commands were confused by the dep overrides being elsewhere.

Is there a sense of what needs to be done for Melos and Pub workspaces to coexist?

**Comment by spydon** [comment]
@xaethos it shouldn't be a problem for them to co-exist already afaik. With the workspaces we can simplify a lot of our code in Melos and remove the packages config from melos.yaml though, and that needs some work.

**Comment by Leptopoda** [comment]
What I have done so far:
- remove dep overrides
- add every package to the workspace in the root pubspec
- change the dependency resolution for every package to "workspace"

Et voilà.
You MUST NOT use `melos bootstrap` as it will try to generate the overrides again. I suggest to just use `dart pub get` and melos for all the other commands (format, run scripts, publish, ...).

The current stable flutter version is broken for workspaces and last time I checked master it was also not working. I found `3.26.0-0.1.pre`to be working in some testing (just specify it in fvm) although your use case might forbid unreleased versions. A stable release should be right around the corner.

**Comment by SAGARSURI** [comment]
finally workspace is released to stable version 3.27:

https://dart.dev/tools/pub/workspaces

**Comment by SAGARSURI** [comment]
I hope `melos bootstrap` will not be removed but tweaked to handle the new workspace changes.

**Comment by spydon** [comment]
> I hope `melos bootstrap` will not be removed but tweaked to handle the new workspace changes.
> 
>     * This way we can still use `hooks` pre/post `bootstrap` command execution.
> 
>     * Also the variable `MELOS_PACKAGES` might get effected which we use it in our pipeline to run test on changed packages because of this problem [fix:  melos overrides the applied filter for dependents/dependencies #675](https://github.com/invertase/melos/issues/675)

Yeah, that's true. We should keep bootstrap, but as an optional step.

I'm pretty busy currently, so if there is anyone else that would like to be assigned to this issue, give me a ping!
I don't think it is thaaat much to do, mostly removing a lot of stuff that we have in melos that is now superfluous.

**Comment by lishaduck** [comment]
If anyone is looking for a working example of a migration, I've got this PR: lishaduck/legacy_checks#38.

**Comment by SalihCanBinboga** [comment]
> If anyone is looking for a working example of a migration, I've got this PR: [lishaduck/legacy_checks#38](https://github.com/lishaduck/legacy_checks/pull/38).

“melos bootstrap” command cannot be executed by you, right? (“melos bootstrap” command creates a ‘pubspec_overrides.yaml’ file for each module. Because of this, pub workspaces do not function properly)

**Comment by lishaduck** [comment]
> “melos bootstrap” command cannot be executed by you, right? (“melos bootstrap” command creates a ‘pubspec_overrides.yaml’ file for each module. Because of this, pub workspaces do not function properly)

Right. You call `dart pub get` from the root instead. See here: https://github.com/lishaduck/legacy_checks/pull/38#discussion_r1901916407.

**Comment by SalihCanBinboga** [comment]
> > “melos bootstrap” command cannot be executed by you, right? (“melos bootstrap” command creates a ‘pubspec_overrides.yaml’ file for each module. Because of this, pub workspaces do not function properly)
> 
> Right. You call `dart pub get` from the root instead. See here: [lishaduck/legacy_checks#38 (comment)](https://github.com/lishaduck/legacy_checks/pull/38#discussion_r1901916407).

Thanks @lishaduck  but I used the 'melos bootstrap' command because it also generates *.iml files specific to modules in IntelliJ. Additionally, when I run build_runner in the root directory, it still doesn't detect the modules - I need to go to the relevant module directory and run build_runner there. Hopefully this will be resolved soon..

**Comment by SalihCanBinboga** [comment]
> > Thanks @lishaduck but I used the 'melos bootstrap' command because it also generates *.iml files specific to modules in IntelliJ.
> 
> Yeah, I use VS Code, which doesn't need them. What'd I'd probably do is run bootstrap, run `rm -rf .dart_tool **/pubspec_overrides.yaml`, and then run `dart pub get`.
> 
> > Additionally, when I run build_runner in the root directory, it still doesn't detect the modules - I need to go to the relevant module directory and run build_runner there. Hopefully this will be resolved soon..
> 
> I'd think that's the expected behavior? Just make a melos task to run it all concurrently.

Actually, I already have the melos run command as shown below, but due to bootstrap pub workspaces, this command cannot be executed, so it doesn't apply to all modules. I'm in a dilemma in such a situation :)

`  gen:all:
    name: gen
    description: Generate all files
    run: "dart run build_runner build -d --low-resources-mode"
    packageFilters:
      dependsOn:
        - "build_runner"
    exec:
      concurrency: 4
      failFast: true`

**Comment by lishaduck** [comment]
> Actually, I already have the melos run command as shown below, but due to bootstrap pub workspaces, this command cannot be executed, so it doesn't apply to all modules. I'm in a dilemma in such a situation :)

What's the error? (I haven't migrated my `build_runner`-dependent packages because `custom_lint` support is still in-progress)

**Comment by SalihCanBinboga** [comment]
> > Actually, I already have the melos run command as shown below, but due to bootstrap pub workspaces, this command cannot be executed, so it doesn't apply to all modules. I'm in a dilemma in such a situation :)
> 
> What's the error? (I haven't migrated my `build_runner`-dependent packages because `custom_lint` support is still in-progress)

Problem: Due to the issue mentioned [here](https://github.com/invertase/melos/issues/747#issuecomment-2569114225), Melos does not recognize my modules, so I cannot run my Melos scripts for my modules.

**Comment by lishaduck** [comment]
> Problem: Due to the issue mentioned [here](https://github.com/invertase/melos/issues/747#issuecomment-2569114225), Melos does not recognize my modules, so I cannot run my Melos scripts for my modules.

Can you just copy+paste the error message? That message doesn't provide an error nor does it mention Melos failing to recognize modules.

**Comment by SalihCanBinboga** [comment]
> > Problem: Due to the issue mentioned [here](https://github.com/invertase/melos/issues/747#issuecomment-2569114225), Melos does not recognize my modules, so I cannot run my Melos scripts for my modules.
> 
> Can you just copy+paste the error message? That message doesn't provide an error nor does it mention Melos failing to recognize modules.

Melos needs to recognize the modules, which requires running the `melos bootstrap` method, right? Since I cannot run the `melos bootstrap` method due to the use of the pub workspace, the modules are not recognized. Therefore, when I run `melos gen:all`, it does not give any errors, but it does not recognize any of the 60 modules.

**Comment by lishaduck** [comment]
> Melos needs to recognize the modules, which requires running the `melos bootstrap` method, right?

I don't think so... it just uses a glob in the melos.yaml, right?

**Comment by spydon** [comment]
> I don't think so... it just uses a glob in the melos.yaml, right?

Correct, there is no caching of the package list when bootstrapping.

**Comment by SalihCanBinboga** [comment]
@lishaduck @spydon You are right; I noticed the issue was different. After enabling pub workspaces, I removed the build_runner dependency from the modules because it is sufficient to use it only in the root workspace. I realized that the gen:all script doesn’t recognize the modules because it depends on build_runner..

178 directories

**Comment by spydon** [comment]
The PR to migrate Melos to Pub Workspaces is now out, feel free to review it!
https://github.com/invertase/melos/pull/816

Tomorrow I'll create a pre-release so that you can try it out in your projects.

**Comment by spydon** [comment]
If you want to add `resolution: workspace` to all of your `pubspec.yaml` files you can do this:
```sh
find . -type f -name "pubspec.yaml" -exec sed -i '/^name:/a resolution: workspace' {} +
```
and to get all your `pubspec.yaml` paths to add to the root pubspec's `workspace` section, you can do this:
```sh
find . -name pubspec.yaml
```

Remember that the `workspace` section doesn't support globs yet, so you'll most likely have quite a long list, you can give a thumbs up to the issue suggesting to add globs to it here: https://github.com/dart-lang/pub/issues/4391

**Comment by spydon** [comment]
If anyone is curious how a migration might look like, this is what it looked like for the Flame monorepo:
https://github.com/flame-engine/flame/pull/3433

**Comment by lishaduck** [comment]
Oh, and in case anyone is using `custom_lint`: don't upgrade 😢
`custom_lint` is essentially unmaintained right now, as the Dart team is actively working on shipping analyzer plugins built-in... which means Remi isn't going to put in the effort to get `custom_lint` working in Pub Workspaces.
Alternatively, turn off `custom_lint`.

**Comment by spydon** [comment]
@lishaduck there was a release for `custom_lint` two days ago, was that one not working with workspaces either?

**Comment by lishaduck** [comment]
> @lishaduck there was a release for `custom_lint` two days ago, was that one not working with workspaces either?

I haven't tested it, but it shouldn't've fixed it. He bumped `analyzer`, `source_gen`, etc, across his ecosystem (Riverpod, Freezed, custom_lint, etc), but workspace support wasn't included, and, as I noted, isn't planned (invertase/dart_custom_lint#298).

**Comment by petrnymsa** [comment]
Probably not related directly with melos but until Pub workspaces were introduced we used melos even in projects which are not monorepos. Why? Because every project has several "standartized" melos scripts such as setup, codegen.... 

However after Melos is migrated to support pub workspaces and removed melos.yaml we cant do that anymore as `- .` relative path in workspace setting is not allowed. 

```yaml
# Main pubpsec.yaml

workspace:
	- . 

.....

melos:
  scripts:
      ....

```

Executing `melos bs` 

```cli

fvm exec melos bs 
fvm: Running version: "3.27.1"

Error on line 10, column 5 of pubspec.yaml: "workspace" members must be subdirectories
   ╷
10 │   - .
   │     ^
   ╵
```

**Comment by abikko** [comment]
@spydon Hello, thanks for the great package 🚀  

Is that means that 7.0.0 melos is stable?

> `7.0.0-dev.3` is now out, removing the root `melos.yaml` and moving the config in under a melos section in `pubspec.yaml` since half of the config is in there now anyways, migration instructions are updated in the readme and in the docs.
> 
> I'll keep this issue open until we have released 7.0.0 to stable.
> 
> Do note that if you have `uses-material-design: true` in an example and not in the package itself and runs `flutter test` in the package you will get a warning, but it will soon be fixed. ([dart-lang/pub#4486](https://github.com/dart-lang/pub/issues/4486))

**Comment by petrnymsa** [comment]
@spydon  Not sure (again) if I should create new issue for this. 

After we migrated single app project to newer melos (see previous comment) we cant run melos command inside "app" folder. Melos needs to be executed in root folder, where workspace pubspec.yaml is defined along with pubspec.lock. Otherwise melos looks for pubspec.lock inside app folder. 

```
/root
  - pubpsec.yaml // WORKSPACE pubspec
  - pubspec.lock
  - app/
        - pubspec.yaml // APP pubspec
```

If I run melos run inside app folder I am getting 
```
PathNotFoundException: Cannot open file, path = 'REDACTED/app/pubspec.lock' (OS Error: No such file or directory, errno = 2)
#0      _File.throwIfError (dart:io/file_impl.dart:675:7)
#1      _File.openSync (dart:io/file_impl.dart:490:5)
#2      _File.readAsBytesSync (dart:io/file_impl.dart:574:18)
#3      _File.readAsStringSync (dart:io/file_impl.dart:624:18)
#4      ExecutableInstallation._loadPubspecLockEntry (package:cli_launcher/cli_launcher.dart:142:47)
#5      ExecutableInstallation._pubspecLockEntry (package:cli_launcher/cli_launcher.dart:138:34)
#6      ExecutableInstallation._pubspecLockEntry (package:cli_launcher/cli_launcher.dart)
#7      ExecutableInstallation._loadIsFromPath (package:cli_launcher/cli_launcher.dart:169:19)
#8      ExecutableInstallation.isFromPath (package:cli_launcher/cli_launcher.dart:127:42)
#9      launchExecutable (package:cli_launcher/cli_launcher.dart:396:29)
```

**Comment by spydon** [comment]
@petrnymsa hmm, I can't reproduce this.
Could you create a minimal reproducible example in a separate repo?

**Comment by petrnymsa** [comment]
> [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/21087414?s=64&amp;u=00c32435d2f481bd01ba776fc8d512e8deeabfed&amp;v=4">@petrnymsa](https://github.com/petrnymsa) hmm, I can't reproduce this. Could you create a minimal reproducible example in a separate repo?

Hmm me neither in sandbox project. We are using FVM as well, maybe something is off there

**Comment by spydon** [comment]
@petrnymsa what about if you add FVM to the MRE, is it reproducible then?

**Comment by petrnymsa** [comment]
@spydon  unfortunately no, will let you know If I find a way to reproduce it

**Comment by petrnymsa** [comment]
@spydon  Got it. During migration, we forgot melos inside dev_dependcies in app/pubspec.yaml. After removal, everything works.

**Comment by markbeij** [comment]
I got [7.0.0-dev.4](https://pub.dev/packages/melos/versions/7.0.0-dev.4) working. What work needs to be done before going stable?

**Comment by spydon** [comment]
> I got [7.0.0-dev.4](https://pub.dev/packages/melos/versions/7.0.0-dev.4) working. What work needs to be done before going stable?

The shared dependencies feature needs some love, because if you set dependencies that don't work in there you won't be able to run it again to fix it, so it needs some kind of rollback feature.

Other than that everything looks good so far!

**Comment by markbeij** [comment]
Thanks Lukes. Is there already a ticket for the required love?

From: Lukas Klingsbo ***@***.***>
Date: Wednesday, 22 January 2025 at 09:40
To: invertase/melos ***@***.***>
Cc: Mark Beij ***@***.***>, Manual ***@***.***>
Subject: Re: [invertase/melos] request: pub workspaces (Issue #747)

I got 7.0.0-dev.4<https://pub.dev/packages/melos/versions/7.0.0-dev.4> working. What work needs to be done before going stable?

The shared dependencies feature needs some love, because if you set dependencies that don't work in there you won't be able to run it again to fix it, so it needs some kind of rollback feature.

Other than that everything looks good so far!

**Comment by spydon** [comment]
@markbeij created one here now: 
https://github.com/invertase/melos/issues/847

**Comment by spydon** [comment]
7.0.0-dev.5 is now out and it has the shared dependencies rollback feature included.

**Comment by spydon** [comment]
I think we are very soon ready for a stable release now, I'm just waiting for [this one](https://github.com/flutter/flutter/pull/160443) to land and be released.

It would be nice with glob support in the `workspace` list too, but I think that might take a bit longer (https://github.com/dart-lang/pub/issues/4391). Give it a thumbs up if that would be a useful feature for your too.

**Comment by XavierChanth** [comment]
I know a vast majority of people will prefer workspaces. But for the exact reason that people prefer it, it's actually caused problems for my use case.

We have a whole ecosystem of packages which are in separate repos by type/purpose (a bunch of categorized monorepos if you will).

A rough summary:
- core dart libraries
- server packages & binaries
- client sdk packages
- flutter packages
- services & products built on the platform

We have an additional repo called [at_mono](https://github.com/atsign-foundation/at_mono), where we are using melos as a tool to consolidate git submodules of these various repos in one place. We then do dependency analysis and some other testing, including multi-language compatibility tests.

Melos 6 used a distributed package versions to centralized resolution strategy, which fit this use case very well. 

Melos 7, on the other hand, has an inverse approach where package versions start centralized in the workspace then are distributed for resolution. For the vast majority of people, its obvious why this is a better option, package versions remain compatible since they are centralized.

I have a reasonable strategy for migrating using some pre-bootstrap scripts, but nonetheless, wanted to share some of the realized limitations of melos moving to workspaces.

**Comment by spydon** [comment]
@XavierChanth yeah, there are not only upside to the new workspace feature, but for most projects the upsides outweigh the downsides.

**Comment by lishaduck** [comment]
For those of us like me waiting custom_lint support, a PR was raised yesterday, and it looks like the CI failure is a flaky test that timed out, so hopefully it'll get merged soon: https://github.com/invertase/dart_custom_lint/pull/312.

**Comment by lishaduck** [comment]
custom_lint landed, melos+pub workspaces+custom_lint+codegen all works together beautifully. And opening my project doesn't tank my computer anymore too, which is nice :)

**Comment by markbeij** [comment]
Would love to get an update. Curious what's the bottleneck at the moment
Can't wait to have this feature 😄

**Comment by spydon** [comment]
> Would love to get an update. Curious what's the bottleneck at the moment
> Can't wait to have this feature 😄 

It is already available in v7.0.0-dev.7 :)
As mentioned above, we're waiting for https://github.com/flutter/flutter/pull/160443 to be merged and released before we'll make it stable.

**Comment by youssefali424** [comment]
Question regarding Workspaces migration => what to do about apps folder ? @spydon 
Apps should be a workspace resolution ? as this is the only option right as it's not possible to have package overrides and this isn't mentioned in the docs

**Comment by lishaduck** [comment]
Huh? Just replace wherever you see `packages/` with `apps/`. Pub doesn't care.

**Comment by youssefali424** [comment]
@lishaduck may be you didn't get what i meant, i was referring to the docs [here](https://melos.invertase.dev/getting-started#install-melos-in-the-workspace) it clearly mention that the expected workspaces are the packages folder not the apps, btw i already have a working example the same as you said by just making the apps inside the workspace structure as well (resolution = workspace with the apps as well), but the question is that is the correct integration or not

**Comment by lishaduck** [comment]
> is that the correct integration or not ?

Yes, that's what I'm saying. Melos recommends using `packages/` for simple monorepos, but remains agnostic.

**Comment by spydon** [comment]
@lishaduck is correct, the directories can be named anything, `packages` and `apps` are just recommendations.

**Comment by spydon** [comment]
The PR with all the fixes is merged, but not released on stable yet:
https://ismyprinflutteryet.com/#/pr/framework/160443

If we're lucky we'll get it in a hotfix release.

**Comment by Him-Sharma** [comment]
We are integrating our flutter repo as a module in existing android and iOS projects, which relies on .ios and .android dirs to be created. When we use dart workspaces and use melos bootstrap, it is not running flutter pub get in the workspace dirs, hence the required .ios and .android dirs are not created. 
anyone else faced this issue and knows how to work around this?

**Comment by spydon** [comment]
> We are integrating our flutter repo as a module in existing android and iOS projects, which relies on .ios and .android dirs to be created. When we use dart workspaces and use melos bootstrap, it is not running flutter pub get in the workspace dirs, hence the required .ios and .android dirs are not created.

That sounds like a bug that should be reported on Flutter's issue tracker, since I'm guessing this happens without Melos too?

**Comment by Him-Sharma** [comment]
> > We are integrating our flutter repo as a module in existing android and iOS projects, which relies on .ios and .android dirs to be created. When we use dart workspaces and use melos bootstrap, it is not running flutter pub get in the workspace dirs, hence the required .ios and .android dirs are not created.
> 
> That sounds like a bug that should be reported on Flutter's issue tracker, since I'm guessing this happens without Melos too?

I was also thinking about that but as a user of melos bootstrap command, I felt this was a regression in behavior. let me raise this in flutter repo 

(Edit) found this issue in flutter already https://github.com/flutter/flutter/issues/161927

**Comment by markbeij** [comment]
Do I understand correctly that we have to wait for [flutter/flutter#161927](https://github.com/flutter/flutter/issues/161927) to be in stable before this issue can be finished?
Currently it's only in master, see https://ismyprinflutteryet.com/#/pr/framework/170517

**Comment by peter-trost** [comment]
Edit: Melos 7.0.0 with pub workspaces support has already been released 5 hours ago 😂
Thanks a lot @spydon 🙏

@spydon I think the fix is now merged and released with 3.35.0 🚀

[Source](https://github.com/flutter/flutter/pull/170517)
<img width="870" height="88" alt="Image" src="https://github.com/user-attachments/assets/0e4e2ccb-6e6f-4cad-85c2-818ee3f914e0" />

**Comment by markbeij** [comment]
[celebrate]     Mark Beij reacted to your message:
________________________________

**Comment by lishaduck** [comment]
> [@spydon](https://github.com/spydon) I think the fix is now merged and released with 3.35.0 🚀

Beat me to it, I was quite literally loading the page to post that lol
I kept meaning to get to it yesterday.

**Comment by spydon** [comment]
Yes, fiiiinnally! 🥳🥳
